test/e2e_node/swap_test.go

```diff
@@ -38,49 +45,231 @@ import (
+			ginkgo.By(fmt.Sprintf("Creating a pod of QOS class %s. memoryRequestEqualLimit: %t", qosClass, memoryRequestEqualLimit))
+			pod := getSwapTestPod(f, qosClass, memoryRequestEqualLimit)
+			pod = runPodAndWaitUntilScheduled(f, pod)
+
+			gomega.Expect(isPodCgroupV2(f, pod)).To(gomega.BeTrueBecause("cgroup v2 is required for swap"))
+			gomega.Expect(getSwapBehavior()).To(gomega.Or(gomega.Equal(types.NoSwap), gomega.BeEmpty()), "NodeConformance is expected to run with NoSwap")
+
+			expectNoSwap(f, pod)
+		},
+			ginkgo.Entry("QOS Best-effort", v1.PodQOSBestEffort, false),
+			ginkgo.Entry("QOS Burstable", v1.PodQOSBurstable, false),
+			ginkgo.Entry("QOS Burstable with memory request equals to limit", v1.PodQOSBurstable, true),
+			ginkgo.Entry("QOS Guaranteed", v1.PodQOSGuaranteed, false),
+		)
+	})
 
-		if !isLimitedSwap {
-			ginkgo.By("expecting no swap")
-			expectNoSwap(f, pod, isCgroupV2)
-			return
+	f.Context(framework.WithSerial(), func() {
+
+		enableLimitedSwap := func(ctx context.Context, initialConfig *config.KubeletConfiguration) {
+			msg := "swap behavior is already set to LimitedSwap"
+
+			if swapBehavior := initialConfig.MemorySwap.SwapBehavior; swapBehavior != types.LimitedSwap {
+				initialConfig.MemorySwap.SwapBehavior = types.LimitedSwap
+				msg = "setting swap behavior to LimitedSwap"
+			}
+
+			ginkgo.By(msg)
 		}
 
-		ginkgo.By("expecting limited swap")
-		expectedSwapLimit := calcSwapForBurstablePod(f, pod)
-		expectLimitedSwap(f, pod, expectedSwapLimit)
-	},
-		ginkgo.Entry("QOS Best-effort", v1.PodQOSBestEffort, false),
-		ginkgo.Entry("QOS Burstable", v1.PodQOSBurstable, false),
-		ginkgo.Entry("QOS Burstable with memory request equals to limit", v1.PodQOSBurstable, true),
-		ginkgo.Entry("QOS Guaranteed", v1.PodQOSGuaranteed, false),
-	)
+		f.Context("Basic functionality", func() {
+			tempSetCurrentKubeletConfig(f, enableLimitedSwap)
+
+			ginkgo.DescribeTable("with configuration", func(qosClass v1.PodQOSClass, memoryRequestEqualLimit bool) {
+				ginkgo.By(fmt.Sprintf("Creating a pod of QOS class %s. memoryRequestEqualLimit: %t", qosClass, memoryRequestEqualLimit))
+				pod := getSwapTestPod(f, qosClass, memoryRequestEqualLimit)
+				pod = runPodAndWaitUntilScheduled(f, pod)
+
+				gomega.Expect(isPodCgroupV2(f, pod)).To(gomega.BeTrueBecause("cgroup v2 is required for swap"))
+				gomega.Expect(getSwapBehavior()).To(gomega.Equal(types.LimitedSwap))
+
+				expectedSwapLimit := calcSwapForBurstablePod(f, pod)
+				expectLimitedSwap(f, pod, expectedSwapLimit)
+			},
+				ginkgo.Entry("QOS Best-effort", v1.PodQOSBestEffort, false),
+				ginkgo.Entry("QOS Burstable", v1.PodQOSBurstable, false),
+				ginkgo.Entry("QOS Burstable with memory request equals to limit", v1.PodQOSBurstable, true),
+				ginkgo.Entry("QOS Guaranteed", v1.PodQOSGuaranteed, false),
+			)
+		})
+
+		// These tests assume the following, and will fail otherwise:
```

Comment:
thank you for those. What swappiness value did you use in your configs? is it something we need to take into consideration. I understand that kernel may not be super efficient in swapping some times. So depending on how fast we write memory, test may start flaking.

The soft memory limit in cgroupv2 is something that I think may improve swap reliability.

Reply:
>thank you for those. What swappiness value did you use in your configs? is it something we need to take into consideration.

Thanks for your review Sergey!

I've spent a lot of time writing these tests in a way that they will be as non-flaky and stable as possible, but at the same time will verify that every expectation from the NodeSwap feature is met. I've followed a few principles in order to achieve it:
* **The system is not configured in any special way**: The tests should work with default OS configurations to prove that NodeSwap is not dependent on any specific configuration like swappiness.
* **Tests end once the most basic functionality is verified**: For example, when ensuring that a container can exceed memory limits, the container (slowly but surely) stresses memory and once `memory.current > limits` and `memory.swap.current > 0` the test passes.

> So depending on how fast we write memory, test may start flaking.

Bear in mind that the stress is being made at a low pace to give the kernel enough time to reclaim. It is proven to be very stable until now, but we can always make it even slower if we'll see future flakiness.

I think that swappiness shouldn't matter too much here as well. Essentially it is a way to tell the kernel to prefer reclaiming file-backed memory vs anon memory, but in any case after enough stress the kernel would have to start swapping.

> The soft memory limit in cgroupv2 is something that I think may improve swap reliability.

I think you're referring to `memory.swap.high`.
This is indeed an interesting topic which I think about a lot. I'm not sure if it's in scope for this PR, but I would love discussing it offline if you have the time/interest.